config/initializers/0_application_config.rb

```diff
@@ -4,4 +4,10 @@ class ApplicationConfig
   def self.[](key)
     ENVied.send(key)
   end
+
+  def self.app_domain_no_port
+    app_domain = self["APP_DOMAIN"]
+    match = app_domain.match(/^(?<host>.+?)(?<port>:\d+)?$/)
```

Comment:
Sometimes I really wish `URI.regexp` had proper named capture groups, since it implements the full scheme and is much more robust than anything we're likely to write ourselves [1]. Alas it doesn't and the capture group ordering is unstable, so it's pretty useless for any purpose other than validation.

There's a potential problem with this though: if you want to allow e.g. an ngrok URL, the protocol will be matched as part of the host, which is not what we want in this case:

```ruby
md = "http://test123.ngrok.io".match(/^(?<host>.+?)(?<port>:\d+)?$/)
md["host"]
#=> "http://test123.ngrok.io"
```

Maybe this would be a bit better:

```ruby
# No need to recompile the regex on every method invocation
URI_REGEXP = %r{(?<scheme>https?://)?(?<host>.+?)(?<port>:\d+)?$}.freeze

def self.app_domain_no_port
  app_domain = self["APP_DOMAIN"]
  match = app_domain.match(URI_REGEXP)
  match[:host]
end
```

The regex seems to work well enough in quick tests:

```ruby
"http://test123.ngrok.io".match(URI_REGEXP)
#=> #<MatchData "http://test123.ngrok.io" scheme:"http://" host:"test123.ngrok.io" port:nil>
"https://test123.ngrok.io".match(URI_REGEXP)
#=> #<MatchData "https://test123.ngrok.io" scheme:"https://" host:"test123.ngrok.io" port:nil>
"localhost:3000".match(URI_REGEXP)
#=> #<MatchData "localhost:3000" scheme:nil host:"localhost" port:":3000">
```

[1] Here's the regular expression in all its glory:

```ruby
/
        ([a-zA-Z][\-+.a-zA-Z\d]*):                           (?# 1: scheme)
        (?:
           ((?:[\-_.!~*'()a-zA-Z\d;?:@&=+$,]|%[a-fA-F\d]{2})(?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*)                    (?# 2: opaque)
        |
           (?:(?:
             \/\/(?:
                 (?:(?:((?:[\-_.!~*'()a-zA-Z\d;:&=+$,]|%[a-fA-F\d]{2})*)@)?        (?# 3: userinfo)
                   (?:((?:(?:[a-zA-Z0-9\-.]|%\h\h)+|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|\[(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|(?:(?:[a-fA-F\d]{1,4}:)*[a-fA-F\d]{1,4})?::(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}))?)\]))(?::(\d*))?))? (?# 4: host, 5: port)
               |
                 ((?:[\-_.!~*'()a-zA-Z\d$,;:@&=+]|%[a-fA-F\d]{2})+)                 (?# 6: registry)
               )
             |
             (?!\/\/))                           (?# XXX: '\/\/' is the mark for hostport)
             (\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*(?:\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*)*)?                    (?# 7: path)
           )(?:\?((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                 (?# 8: query)
        )
        (?:\#((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                  (?# 9: fragment)
      /x
```

Reply:
@citizen428 thanks for the tip about `URI`. It actually has `#host`, `#port`, etc. I was hopeful, but it doesn't work for schemaless domains:
```
[1] pry(main)> URI("renner").host
=> nil
[1] pry(main)> URI("http://renner").host
=> "renner"
```

Thanks a lot for writing code, and optimising regex compilation. I've used that, and added tests.